File: src/game/rules/general_rules.lua
local GeneralRules = {}
GeneralRules.__index = GeneralRules

function GeneralRules:new(player_count, shared_board, starting_hand_size, maximum_hand_size, can_exceed_max_hand_size, win_conditions, response_mode)
    local generalRules = setmetatable({}, GeneralRules)

    generalRules.player_count = player_count or 2
    generalRules.shared_board = shared_board or false
    generalRules.starting_hand_size = starting_hand_size or 1
    generalRules.maximum_hand_size = maximum_hand_size
    generalRules.can_exceed_max_hand_size = can_exceed_max_hand_size or true
    generalRules.win_conditions = win_conditions or {}
    generalRules.response_mode = response_mode

    return generalRules
end

return GeneralRules

File: src/server/game/config/general_rules.go
package config

import "sync"

type HandExceedAction struct {
	Name       string                 `json:"name"`
	Parameters map[string]interface{} `json:"parameters"`
}

type CardContainerConfig struct {
	Name   string `json:"name"`
	Public bool   `json:"public"`
}

type GeneralRules struct {
	PlayerCount          []int                 `json:"player_count"`
	SharedBoard          bool                  `json:"shared_board"`
	StartingHandSize     int                   `json:"starting_hand_size"`
	MaximumHandSize      int                   `json:"maximum_hand_size"`
	CanExceedMaxHandSize bool                  `json:"can_exceed_max_hand_size"`
	WinConditions        []string              `json:"win_conditions"`
	ResponseMode         string                `json:"response_mode"`
	HandExceedAction     HandExceedAction      `json:"hand_exceed_action"`
	CardContainers       []CardContainerConfig `json:"card_containers"`
}

type GeneralRulesConfig struct {
	GeneralRules GeneralRules `json:"general_rules"`
}

var (
	generalRulesConfig    sync.Once
	GeneralRulesConfigVar GeneralRulesConfig
)

// GetGeneralRulesConfig allows to get the game configuration linked to the general rules globally.
func GetGeneralRulesConfig() GeneralRulesConfig {
	return GeneralRulesConfigVar
}
